File: lexical.py
def nextPermutation(A):
    A = sorted(list(A))
    swapped = False
    smallest = len(A)-1

    for i in range(len(A)-1, 0 ,-1):

        if(swapped == True):
            break

        if(A[i] < smallest):
            smallest = i

        if(A[i] <= A[i-1]):
            continue

        temp = A[smallest]
        A[smallest] = A[i-1]
        A[i-1] = temp

    A = A[0:i] + list(reversed(A[i:])) 
    swapped = True

    if(swapped == True):
        return A
    else:
        return sorted(A)

# print nextPermutation([ 1,2,3,4 ])
print nextPermutation('cab')

File: lexical.js
function nextPerm(A){
       
    var swapped = false
    var smallest = A.length - 1

    for(var i=A.length-1; i>0;i--){
        if(swapped){
            break;
        }

        if(A[i] < smallest){
            smallest = i;
        }

        if(A[i] <= A[i-1]){
            continue;
        }

        var temp = A[smallest]
        A[smallest] = A[i-1]
        A[i-1] = temp
        
        var t = A.splice(0);
        var t1 = A.splice(0, i);
        var t2 = A.splice(i, A.length).reverse();
        A = t1.concat(t2); 
        swapped = true;
    }       

    if(swapped)
        return A.join('')
    else
        return A.sort().join('')
}

console.log(nextPerm([1,2,3]))